Automattic/woocommerce-payments
# Validate order's status for create customer API.

#### Changes proposed in this Pull Request

This is a continuation from [@htdat's comment](https://github.com/Automattic/woocommerce-payments/pull/2501#pullrequestreview-715645188).

Return 400 error if payments/orders/{order_id}/create_customer API is called for order with status other than `pending`, `processing`, or `on-hold`.

FYI, all possible order statuses are [listed here](https://github.com/woocommerce/woocommerce/blob/trunk/includes/wc-order-functions.php#L95-L101):
- `pending`
- `processing`
- `on-hold`
- `completed`
- `cancelled`
- `refunded`
- `failed`


#### Testing instructions

* Enable Cash of Delivery from wp-admin > WooCommerce > Settings > Payments (/wp-admin/admin.php?page=wc-settings&tab=checkout).
* Create a COD order.
* Change the order status to one of the status other than the allowed ones (`pending`, `processing`, or `on-hold`), eg: `cancelled`.
* Change [this line](https://github.com/Automattic/woocommerce-payments/blob/0381f02e4171a9bafaee4559fbabc6e2356bc4ef/includes/admin/class-wc-rest-payments-orders-controller.php#L75) from:
```
'permission_callback' => [ $this, 'check_permission' ],
```
to
```
'permission_callback' => '__return_true',
```
* Call the API: `curl -X POST {store_url}/wp-json/wc/v3/payments/orders/{order_id}/create_customer`.
* With this PR branch, expect a 400 error.
```json
{"code":"wcpay_invalid_order_status","message":"Invalid order status","data":{"status":400}}
```
* With base branch, expect a success.

-------------------
